data/src/main/java/com/linkedin/data/schema/compatibility/AnnotationCompatibilityChecker.java

```diff
@@ -0,0 +1,193 @@
+  }
+
+  /**
+   * Iterate the nodeToResolverPropertiesMap, if a node's resolvedProperty contains the same annotationNamespace as SchemaAnnotationHandler,
+   * calling annotationCompatibilityCheck api which is provided in the SchemaAnnotationHandler to do the annotation compatibility check.
+   * @param olderResolvedPropertiesMap
+   * @param newerResolvedPropertiesMap
+   * @param handlers SchemaAnnotationHandler list
+   * @return List<AnnotationCompatibilityResult>
+   */
+  static List<AnnotationCompatibilityResult> getCompatibilityResult(Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext,
+      Map<String, Object>>> olderResolvedPropertiesMap, Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext,
+      Map<String, Object>>> newerResolvedPropertiesMap, List<SchemaAnnotationHandler> handlers)
+  {
+    List<AnnotationCompatibilityResult> results = new ArrayList<>();
+    for (SchemaAnnotationHandler handler : handlers)
+    {
+      String annotationNamespace = handler.getAnnotationNamespace();
+      for (Map.Entry<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>> e : olderResolvedPropertiesMap
+          .entrySet())
+      {
+        PathSpec pathSpec = e.getKey();
+        Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>> olderCheckContextAndResolvedProperty =
+            olderResolvedPropertiesMap.get(pathSpec);
+        Map<String, Object> olderResolvedProperties = olderCheckContextAndResolvedProperty.getValue();
+
+        if (newerResolvedPropertiesMap.containsKey(pathSpec))
+        {
+          // If older schema node and newer schema node have the same pathSpec,
+          // they may or may not contain the same annotation namespace as SchemaAnnotationHandler, we need to further check.
+          Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>
+              newerCheckContextAndResolvedProperty = newerResolvedPropertiesMap.get(pathSpec);
+          Map<String, Object> newerResolvedProperties = newerCheckContextAndResolvedProperty.getValue();
+
+          // If olderResolvedProperties or newerResolvedProperties contains the same namespace as the provided SchemaAnnotationHandler,
+          // we do the annotation check
+          if (olderResolvedProperties.containsKey(annotationNamespace) || newerResolvedProperties.containsKey(
+              annotationNamespace))
+          {
+            AnnotationCompatibilityResult result =
+                handler.annotationCompatibilityCheck(olderResolvedProperties, newerResolvedProperties,
+                    olderCheckContextAndResolvedProperty.getKey(), newerCheckContextAndResolvedProperty.getKey());
+            if (!result.isCheckAnnotation())
+            {
+              _log.info("There is no implementation of annotationCompatibilityCheck in SchemaAnnotationHandler: "
+                  + handler.getClass().getName());
+              break;
+            }
```

Comment:
This logic feels odd. Did you consider adding a method on the handler directly?\
 eg,` boolean implementsCompatibilityChecking()`

Reply:
add a method on the handler instead.